cmd/traffic/cmd/manager/internal/mutator/watcher.go

```diff
@@ -389,66 +399,123 @@ func (c *configWatcher) Store(ctx context.Context, ac *agentconfig.Sidecar, upda
+			}
+			create = true
+		}
+
+		if create {
+			cm = &core.ConfigMap{
+				TypeMeta: meta.TypeMeta{
+					Kind:       "ConfigMap",
+					APIVersion: "v1",
+				},
+				ObjectMeta: meta.ObjectMeta{
+					Name:      agentconfig.ConfigMap,
+					Namespace: ns,
+				},
 			}
 		}
+
 		if cm.Data == nil {
 			cm.Data = make(map[string]string)
 		}
-		if cm.Data[ac.AgentName] == yml {
-			// Race condition. Snapshot isn't updated yet, or we wouldn't have gotten here.
-			return nil
-		}
-	}
 
-	if updateSnapshot {
-		c.Lock()
+		// Can't add / remove things to the map at this point.
+		c.configUpdatersMapLock.Lock()
+		// Finishes by deleting the key and then unlocking.
+		defer func() {
+			delete(c.configUpdaters, ns)
+			c.configUpdatersMapLock.Unlock()
+		}()
+
+		cmData := map[string]string{}
+
 		if nm, ok := c.data[ns]; ok {
-			nm[ac.AgentName] = yml
-		}
-		c.Unlock()
-	}
-
-	if create {
-		cm = &core.ConfigMap{
-			TypeMeta: meta.TypeMeta{
-				Kind:       "ConfigMap",
-				APIVersion: "v1",
-			},
-			ObjectMeta: meta.ObjectMeta{
-				Name:      agentconfig.ConfigMap,
-				Namespace: ns,
-			},
-			Data: map[string]string{
-				ac.AgentName: yml,
-			},
-		}
-		dlog.Debugf(ctx, "Creating new ConfigMap %s.%s with %s", agentconfig.ConfigMap, ns, ac.AgentName)
-		_, err = api.Create(ctx, cm, meta.CreateOptions{})
-	} else {
-		if _, ok := cm.Data[ac.AgentName]; ok {
-			dlog.Debugf(ctx, "Updating %s in ConfigMap %s.%s", ac.AgentName, agentconfig.ConfigMap, ns)
+			cmData = copyMap(nm)
+
+			for agentName, yml := range c.configUpdaters[ns].Config {
+				var currAc agentconfig.Sidecar
+				// Ensure that we're not about to overwrite a manually added config entry
+				if err = decode(yml, &currAc); err == nil && currAc.Manual {
+					dlog.Warnf(ctx, "avoided an attempt to overwrite manually added config entry for %s.%s", agentName, ns)
+					continue
+				}
+
+				// Race condition. Snapshot isn't updated yet, or we wouldn't have gotten here.
+				if cm.Data[agentName] == yml {
+					continue
+				}
+
+				cmData[agentName] = yml
+
+				if _, updateSnapshotOK := c.configUpdaters[ns].AddToSnapshot[agentName]; updateSnapshotOK {
+					c.Lock()
+					nm[agentName] = yml
+					c.Unlock()
+				}
+
+			}
+		}
+
+		if create {
+			cm.Data = cmData
+			dlog.Debugf(ctx, "Creating new ConfigMap %s.%s", agentconfig.ConfigMap, ns)
+			_, err = api.Create(ctx, cm, meta.CreateOptions{})
 		} else {
-			dlog.Debugf(ctx, "Adding %s to ConfigMap %s.%s", ac.AgentName, agentconfig.ConfigMap, ns)
+			dlog.Debugf(ctx, "Updating ConfigMap %s.%s", agentconfig.ConfigMap, ns)
+			cm.Data = cmData
+			_, err = api.Update(ctx, cm, meta.UpdateOptions{})
 		}
-		cm.Data[ac.AgentName] = yml
-		_, err = api.Update(ctx, cm, meta.UpdateOptions{})
+
+		return err
 	}
-	return err
+}
+
+func copyMap(m map[string]string) map[string]string {
+	mapCopy := make(map[string]string, len(m))
+	for k, v := range m {
+		mapCopy[k] = v
+	}
+	return mapCopy
 }
```

Comment:
Remove this in favor of [maps](https://github.com/telepresenceio/telepresence/tree/release/v2/pkg/maps).